### DISCUSSION TITLE
Can I get patches between the original tree and the edited one?

### DISCUSSION THREAD

**Discussion by bglgwyng** [tree-sitter/tree-sitter#1460] [Q&A]
Please see the example below.

```javascript
const sourceCode = 'let x = 1; console.log(x);';
const tree = parser.parse(sourceCode);

const newSourceCode = 'const x = 1; console.log(x);';

tree.edit({
  startIndex: 0,
  oldEndIndex: 3,
  newEndIndex: 5,
  startPosition: {row: 0, column: 0},
  oldEndPosition: {row: 0, column: 3},
  newEndPosition: {row: 0, column: 5},
});

const newTree = parser.parse(sourceCode, tree);
```

tree-sitter parses incrementally, so I suppose tree-sitter to get `newTree` by efficiently modifying the original one, `tree`.
Is there any way to get how tree-sitter performed it?
In this example, the patch would be like `replace 'let x = 1;' node by 'const x = 1;' node`.
In more complicated examples, the patch might include many nested replacements, insertions, and deletions of nodes.

**Comment by bglgwyng** [comment]
I found `setLogger` method and I wonder if I can use it for my purpose.
If I can, it's a bit hard to understand how to handle the logs given by `setLogger`. Where can I find documentation for it?

**Comment by alemuller** [comment]
The rust and js files are only used to generate the parser, they are not needed to parse the code at runtime.

The runtime parsing is done by lib-tree-sitter, which is made with the files at `tree-sitter/lib/src`. 

Changes in the code are done by `ts_tree_edit` and `ts_node_edit` functions. The [Using Parsers](https://tree-sitter.github.io/tree-sitter/using-parsers) show how they should be used.

Maybe writing a wrapper for those functions could be used to create a patch?
